### ISSUE TITLE
Adding Xcode build phase using SPM

### ISSUE THREAD

**Issue by IanHoar** [SwiftGen/SwiftGen#689]
Hello 👋

I'm attempting to add SwiftGen using Swift Package Manager and following the integration instructions here: https://github.com/SwiftGen/SwiftGen/blob/master/Documentation/Articles/Xcode-Integration.md

I notice that there isn't documentation on how to access the build script from a SPM installation and I was wondering if there is a preferred path format for locating the installation and calling the script.

I've attempted to use `"$SOURCE_ROOT/SwiftGen/bin/swiftgen"` and am getting warnings about an invalid file location.  

Thanks for your help!

**Comment by crsantos** [comment]
@IanHoar 

From my previous experience with SPM, the previous syntax changed from:

`swift run swiftgen`

To the new one (and starting Xcode 11.4), you need to specify macOS: (it may not be exactly this, so I'll ask you to test):

`/usr/bin/xcrun --sdk macosx swift run swiftgen`

**Comment by djbe** [comment]
I thought SPM didn't support binary dependencies yet. (it's coming in 5.3)

How are you adding it to your project now? Is there something we should add to the documentation?

**Comment by crsantos** [comment]
Maybe a question to @IanHoar , as I'm using Cocoapods for Swiftgen here.

I was just explaining how I call SPM compiled binaries on another projects since Xcode 11.4

**Comment by DenTelezhkin** [comment]
For anyone interested, I'm currently using SwiftGen with SPM following [this article](https://blog.apptekstudios.com/2019/12/spm-xcode-build-tools/)

Couple of things I would like to add:

1. `swift package update` is no longer necessary
2. Automatic package resolution causes issues for some reason, command i'm currently using is 
```bash
swift run --disable-automatic-resolution -c release swiftgen config run --config "$SRCROOT/Configs/.swiftgen.yml"
```
3. Xcode input-ouput file system is very smart, and allows to not run a tool if files were not changed, but you need to add all .stencil files you are using, resource files used in generation **and** .swiftgen.yml config to input files, and all generated files in output files.
4. SwiftGen from SPM is slightly slower than from CocoaPods, but not a big deal, when resource file change, it takes about 5-7 seconds to generate code. When they don't change, build phase ends in milliseconds.
5. If you use templates from SwiftGen, you need to specify path manually(templatePath), it will not be able to automatically find it using templateName parameter. All templates are in `BuildTools/.build/checkouts/SwiftGen/templates`
6. I'm using develop branch of SwiftGen, master branch and current release version do not seem to work.

**Comment by Cyberbeni** [comment]
@DenTelezhkin what issues does automatic package resolution cause? I was trying the `--disable-automatic-resolution` option in my GitHub Action that installs swift based tools (either downloads a cached version or builds it from source) and it made the build fail for SwiftLint at the latest version tag (0.40.3) on Linux (it didn't make any difference on macOS).

**Comment by DenTelezhkin** [comment]
@Cyberbeni I don't remember exactly, but i think dependencies were not actually resolving without this flag on Xcode 11(super weird, but it is what it is).

BTW since then there were couple of adjustments in my flow, specifically:

I now use SwiftGen generation for input output files, that shipped in 6.4.0, and also 6.4.0 release instead of develop branch.

**Comment by djbe** [comment]
I've been working on switching our project structure to fully use SPM (see #793), it'll solve some of the workarounds that @DenTelezhkin mentioned. For example, step 5 won't be needed anymore, as SPM now supports resources (and we use it to our advantage). 

The problem is that there's no easy solution to "add a build phase from SPM in your project", as SPM does not have this feature yet. SPM doesn't have the concept of build "tools" yet, and until it does we'll be stuck with these manual steps.

**Comment by scspijker** [comment]
I got this working as a build phase in Xcode 13.3.
- Add SwiftGen as a dependency in Xcode -> Project -> Package Dependencies (Stable branch, and just the executable).
- Removing it as a library from the target under Frameworks, Libraries and Embedded Content.
- Creating a `runSwiftGen.sh` in the project root (because the Build Phase editor has no highlighting and is crappy in pull requests).
- `chmod +x runSwiftGen.sh`
- Adding this script in a build phase:
```
echo "Build Phase - Begin SwiftGen"
./runSwiftGen.sh
echo "Build Phase - End SwiftGen"
```

The contents of _my_ runSwiftGen.sh, change where your SwiftGen.yml is located!

```
#!/bin/sh

echo "runSwiftGen.sh - Running SwiftGen"

cd ${BUILD_DIR%Build/*}SourcePackages/checkouts/SwiftGen

# This is magic
/usr/bin/xcrun --sdk macosx swift run swiftgen config run --config $SRCROOT/swiftgen.yml

# If this stops working, Apple has twisted the system again to break stuff.
# I found this solution by combining these two comments
# https://github.com/SwiftGen/SwiftGen/issues/689#issuecomment-630040692
# https://github.com/SwiftGen/SwiftGen/issues/689#issuecomment-622341803
# Good luck...

echo "runSwiftGen.sh - Done"
```

**Comment by djbe** [comment]
This is now supported using the official SwiftGenPlugin package, which you can find here:
https://github.com/swiftgen/SwiftGenPlugin

Note: it's not "released" yet, it will be once we release 6.6.0

**Comment by fzymek** [comment]
Hey @scspijker, 

I was trying to use your solution but it turned out that Xcode (I'm on 13.4.1) does not see generated files and fails to compile the project. 

What helped me was specifying output files in the build script phase as shown below:
<img width="831" alt="Screenshot 2022-07-27 at 09 10 05" src="https://user-images.githubusercontent.com/1157689/181184428-455f72b8-12c2-4eda-8fe4-55851731d396.png">

I am not sure if this is because in my `swiftgen.yml` I am declaring output files to this directory or it's someting else... 

Here's yml file for reference:

```
strings:
  inputs: MyProject/Resources/Locale/en.lproj
  filter: .+\.strings$
  outputs:
    - templateName: structured-swift5
      output: MyProject/Generated/Strings.swift
```

**Comment by djbe** [comment]
Implemented by SwiftGen 6.6.0, which has been released. We now have an official [SwiftGenPlugin](https://github.com/SwiftGen/SwiftGenPlugin), where you can find all the relevant information.
